### PR TITLE
Bugfix 549

### PR DESCRIPTION
**Pull request by ThisMatt**
## Description

Allows direct rule transitions that have no changes to be considered valid.
Prevents modification of board elements that would affect existing future case rules.

Specifics:
- model/rules/DirectRule.java
  - Adds a case allowing transitions with no modified data
- model/tree/TreeTransition.java.propogateChange()
  - Propagation now updates modified data arrays of child transitions
  - Other minor fixes
- model/rules/CaseRule.java
  - Added function dependentElements(), which specifies elements to be made unmodifiable
    - Added default implementation and overrides for case rules I thought needed them
- model/gameboard/PuzzleElement.java
  - Added .caseDepended attribute, which counts the number of case rules depending on this element
- model/tree/TreeTransition.java
  - Added .selection attribute, which, for case rule transitions, tracks the selected puzzle element
    - This is neccessary to call dependentElements() given only a transition
    - Side Note: this should help in streamlining checkRule() implementations for case rules
      - Replaced skyscrapers-specific implementation of this idea with the new attribute
        -  CellForNumberCaseRule.java, SkyscrapersClue.java, SkyscrapersBoard.java
 - history/AutoCaseRuleCommand.java
   - Set selection attribute of the created transitions
 - proofeditorui/treeview/TreeView.java
   - Added code to update casesDepended for all puzzleElements in the tree whenever:
     - The first transition of a case rule is added
     - The last transition of a case rule is removed
     - A node containing a case rule is added
     - A node containing a case rule is removed
   - If the Case has been added
     -  All dependentElements in all ancestors are incremented and set to unmodifiable
   - If the Case has been removed
     - All dependentElements in all ancestors are decremented
     - If dependentElements is now 0, sets modifiability to what it should be:
       - Modifiable if started modifiable and not modified by any ancestors

Closes #(549)

## Type of change

- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (improvement to an already existing feature)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Test using Lightup.SatisfyNumber Case Rule:

1. Open 7x7 Easy 69495
2. Apply Finish With Empty to the top-left 0 
    

3. Edit this transition, and notice that when removing all changes, the transition is still valid 

4. Apply case rule to the left-most 2. Note that transition 2 cannot be modified anywhere placing a light would light up a square adjacent to the left-most 2, invalidating a case. (empty placed where possible in this screenshot)
    

5. From the node shown after transition 4, call Finish With Empty again, filling in the spots now unmodifiable by transition 2. Note that all cells not previously modified are modifiable.
    

6. From node 5, Apply case rule to the 2 diagonal to the previous. Note that transition 5 is unmodifiable where placing a light would affect the new 2, and transition 2 cannot be modified where it would affect either 2. (light placed where possible in these screenshots)
  
     

7. Delete each transition after node 5 individually. Note that modifiability remains the same until the final transition is deleted, at which point we return to the state after step 5.
    

8. Undo these deletions. Note that after the first undo, modifiability returns to the state after step 6.
9. Delete node 2. Note modifiability of transition 2 is the same after step 2.
    

10. Undo this deletion. Note modifiability returns to the state after step 6.

## Checklist:

- [x ] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by charlestian23** [comment]
@ThisMatt Can you resolve the merge conflicts? I'll review it once you've done that.

File: src/main/java/edu/rpi/legup/puzzle/treetent/rules/LinkTreeCaseRule.java [hunk 1]
```diff
@@ -152,4 +153,42 @@ public String checkRuleRaw(TreeTransition transition) {
...
+     * Defaults to any element modified by any case
+     */
+    @Override
+    public List<PuzzleElement> dependentElements(Board board, PuzzleElement puzzleElement){
+        List<PuzzleElement> elements = new ArrayList<>(List.of(board.getPuzzleElement(puzzleElement)));
+
+        TreeTentBoard treeTentBoard = (TreeTentBoard) board;
+        TreeTentCell point = (TreeTentCell) puzzleElement;
+
+        //get all adjacent cells
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/puzzle/treetent/rules/LinkTreeCaseRule.java:173]: While is is technically correct, getting each adjacent cell is using duplicate logic. Consider using a `for` loop instead.

Some pseudocode to help:
```java
List<Point> directions = { new Point(0, 1), ..., new Point(-1, 0) };
for (Point direction : directions) {
        TreeTentCell adj = treeTentBoard.getCell(loc.x + point.x, loc.y + point.y);
        ...
        }
```

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java [hunk 2]
```diff
@@ -1,756 +1,881 @@
...
+        }
+        else {
+            TreeTransition trans = (TreeTransition) element;
+            TreeTransitionView transView = (TreeTransitionView) viewMap.get(trans);
+
+            //unlock ancestor elements if case rule deleted
+            Rule rule = trans.getRule();
+            for(TreeNode node : trans.getParents()){
+                //only if the last case of a case rule will be deleted
+                if (rule instanceof CaseRule && node.getChildren().isEmpty()) {
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java:328]: There is an absurd amount of nesting here. They can be reduced by reversing your `if` clauses and then continuing if they're true.

i.e.
```java
if (!(rule instanceof CaseRule && node.getChildren().isEmpty())) {
        continue;
}

CaseRule caseRule = (CaseRule)rule;
...
```

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java [hunk 3]
```diff
@@ -1,756 +1,881 @@
...
+                    //System.out.println("TRANSISTION UNLOCKING");
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java:329]: Remove commented code before submitting changes (even if they're only for debugging).

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java [hunk 4]
```diff
@@ -1,756 +1,881 @@
...
+                                oldElement.setCasesDepended(oldElement.getCasesDepended() - 1);
+                                Point loc = ((GridCell)oldElement).getLocation();
+                                //System.out.println("("+loc.x+","+loc.y+") to "+oldElement.getCasesDepended());
+                                if (oldElement.getCasesDepended() == 0) {
+                                    //set modifiable if started modifiable
+                                    boolean modifiable = tree.getRootNode().getBoard().getPuzzleElement(oldElement).isModifiable();
+
+                                    //unmodifiable if already modified
+                                    TreeNode modNode = ancestor.getParent().getParents().get(0);
+                                    while(modNode.getParent()!=null){
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java:347]: There should be spaces between values and operators.

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java [hunk 5]
```diff
@@ -1,756 +1,881 @@
...
+    public TreeElementView getElementView(TreeElement element) {
+        return viewMap.get(element);
+    }
+
+    private void removeTreeNode(TreeNode node) {
+        viewMap.remove(node);
+        List<TreeTransition> children = node.getChildren();
+
+        //if child is a case rule, unlock ancestor elements
+        if(!children.isEmpty()){
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java:406]: Keep spacing consistent to the format of the existing code. There should be a space between your `if`/`for` keywords, the clause, and the opening brace.

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java [hunk 6]
```diff
@@ -1,756 +1,881 @@
...
+            TreeNodeView nodeView = (TreeNodeView) viewMap.get(node);
+
+            nodeView.getParentView().setChildView(null);
+            removeTreeNode(node);
+        }
+        else {
+            TreeTransition trans = (TreeTransition) element;
+            TreeTransitionView transView = (TreeTransitionView) viewMap.get(trans);
+
+            //unlock ancestor elements if case rule deleted
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java:324]: Space between the comment sequence and the content itself.

```java
//unlock ancestor elements if case rule deleted
```
should be
```java
// unlock ancestor elements if case rule detected
```

File: src/main/java/edu/rpi/legup/puzzle/skyscrapers/SkyscrapersClue.java [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 import edu.rpi.legup.model.gameboard.PuzzleElement;
 
+import java.awt.*;
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/puzzle/skyscrapers/SkyscrapersClue.java:5]: Please refrain from importing whole packages, and only import what you need. Importing more than necessary will affect the time needed to compile and build the project.

File: src/main/java/edu/rpi/legup/puzzle/lightup/rules/SatisfyNumberCaseRule.java [hunk 8]
```diff
@@ -287,4 +287,67 @@ private List<LightUpCell> getAdjacentCells(LightUpBoard board, LightUpCell cell)
...
+     * Returns the elements necessary for the cases returned by getCases(board,puzzleElement) to be valid
+     * Overridden by case rules dependent on more than just the modified data
+     *
+     * @param board         board state at application
+     * @param puzzleElement selected puzzleElement
+     * @return List of puzzle elements (typically cells) this application of the case rule depends upon.
+     * Defaults to any element modified by any case
+     */
+    @Override
+    public List<PuzzleElement> dependentElements(Board board, PuzzleElement puzzleElement){
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/puzzle/lightup/rules/SatisfyNumberCaseRule.java:301]: Space between method header and opening brace.

**Review by Corppet — CHANGES_REQUESTED** [review]

File: src/main/java/edu/rpi/legup/puzzle/treetent/rules/LinkTreeCaseRule.java (hunk 1 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/puzzle/treetent/rules/LinkTreeCaseRule.java]: This was copied from TreeTendBoard.getAdjacent, changed both to a for loop.

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java (hunk 2 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java]: changed

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java (hunk 3 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java]: removed

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java (hunk 4 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java]: added

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java (hunk 5 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java]: added for these and while loops

File: src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java (hunk 6 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java]: added

File: src/main/java/edu/rpi/legup/puzzle/skyscrapers/SkyscrapersClue.java (hunk 7 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/puzzle/skyscrapers/SkyscrapersClue.java]: removed

File: src/main/java/edu/rpi/legup/puzzle/lightup/rules/SatisfyNumberCaseRule.java (hunk 8 above)
Comment (ThisMatt) [reply src/main/java/edu/rpi/legup/puzzle/lightup/rules/SatisfyNumberCaseRule.java]: added

File: src/main/java/edu/rpi/legup/model/rules/CaseRule.java [hunk 9]
```diff
@@ -125,6 +125,31 @@ public String checkRuleAt(TreeTransition transition, PuzzleElement puzzleElement
...
+     * Defaults to any element modified by any case
+     */
+    public List<PuzzleElement> dependentElements(Board board, PuzzleElement puzzleElement) {
+        List<PuzzleElement> elements = new ArrayList<>();
+
+        List<Board> cases = getCases(board,puzzleElement);
+        for (Board caseBoard : cases) {
+            Set<PuzzleElement> data = caseBoard.getModifiedData();
+            for (PuzzleElement element : data) {
+                if(!elements.contains(board.getPuzzleElement(element))){
```
Comment (Corppet) [inline src/main/java/edu/rpi/legup/model/rules/CaseRule.java:145]: Your code still has inconsistent spacing (this comment is only the surface of formatting issues).

**Review by Corppet — CHANGES_REQUESTED** [review]

**Comment by charlestian23** [comment]
I will get around to this after the Thanksgiving break. Also going to request for Chase to review this too, since this seems like a decently substantial change.

**Review by charlestian23 — APPROVED** [review]
Looks good.

**Pull request merged**